sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_interrupt_t uart_irq = (dif_uart_interrupt_t)(
+      (int)plic_irq_id - (int)kTopEarlgreyPlicIrqIdUartTxWatermark);
+  switch (uart_irq) {
+    case kDifUartInterruptTxWatermark:
+      uart_irq_tx_watermark_fired = true;
+      break;
+    case kDifUartInterruptRxWatermark:
+      uart_irq_rx_watermark_fired = true;
+      break;
+    case kDifUartInterruptTxEmpty:
+      uart_irq_tx_empty_fired = true;
+      break;
+    case kDifUartInterruptRxOverflow:
+      uart_irq_rx_overflow_fired = true;
+      break;
+    default:
+      LOG_ERROR("Unsupported unterrupt fired");
+      test_status_set(kTestStatusFailed);
+  }
+
+  // Check if the same interrupt fired at UART as well.
+  dif_uart_enable_t irq_state;
+  dif_uart_irq_state_get(&uart0, uart_irq, &irq_state);
+  CHECK(irq_state == kDifUartEnable,
+        "UART interrupt at fired at PLIC did not fire at UART");
+
+  // Clear the interrupt at UART.
+  dif_uart_irq_state_clear(&uart0, uart_irq);
+
+  // Complete the IRQ at PLIC.
+  dif_plic_irq_complete(&plic0, kTopEarlgreyPlicTargetIbex0, &plic_irq_id);
+}
+
+int main(int argc, char **argv) {
+  test_status_set(kTestStatusInTest);
+
+  // Initialize the UART with:
+  // - device-default clock freq,
+  // - device-default UART baud rate
+  // - parity check disabled.
+  dif_uart_config_t config = {
+      .baudrate = kUartBaudrate,
+      .clk_freq_hz = kClockFreqHz,
+      .parity_enable = kDifUartDisable,
+      .parity = kDifUartParityEven,
+  };
+  mmio_region_t uart_base_addr = mmio_region_from_addr(UART0_BASE_ADDR);
+  bool status = dif_uart_init(uart_base_addr, &config, &uart0);
+  CHECK(status, "dif_uart_init failed");
+  base_set_stdout(uart_stdout);
+
+  // Set the TX and RX watermark to 16 bytes.
+  status = dif_uart_watermark_tx_set(&uart0, kDifUartWatermarkByte16);
+  CHECK(status, "dif_uart_watermark_tx_set failed");
+  status = dif_uart_watermark_rx_set(&uart0, kDifUartWatermarkByte16);
+  CHECK(status, "dif_uart_watermark_rx_set failed");
+
+  // Initialize the PLIC.
+  mmio_region_t plic_base_addr = mmio_region_from_addr(PLIC0_BASE_ADDR);
+  status = dif_plic_init(plic_base_addr, &plic0);
+  CHECK(status, "dif_plic_init failed");
+
+  // Enable these UART interrupts - TX/TX watermark, TX empty and RX overflow.
+  status =
+      dif_uart_irq_enable(&uart0, kDifUartInterruptTxWatermark, kDifUartEnable);
+  status &=
+      dif_uart_irq_enable(&uart0, kDifUartInterruptRxWatermark, kDifUartEnable);
+  status &=
+      dif_uart_irq_enable(&uart0, kDifUartInterruptTxEmpty, kDifUartEnable);
+  status &=
+      dif_uart_irq_enable(&uart0, kDifUartInterruptRxOverflow, kDifUartEnable);
+  CHECK(status, "dif_uart_irq_enable failed");
+
+  // Enable UART interrupts at PLIC as edge triggered.
+  status = dif_plic_irq_trigger_type_set(
+      &plic0, kTopEarlgreyPlicIrqIdUartTxWatermark, kDifPlicEnable);
+  status &= dif_plic_irq_trigger_type_set(
```

Comment:
I think I should probably create an `enum` to have `kDifPlicTriggerTypeLevel/kDifPlicTriggerTypeEdge`. It can be done as a follow up. What do you think?